src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Windows.cs

```diff
@@ -132,7 +133,7 @@ internal static unsafe class AddressChangeListener
             // Callback fired when an address change occurs.
             private static void AddressChangedCallback(object stateObject, bool signaled)
             {
-                lock (s_callerArray)
+                lock (s_globalLock)
```

Comment:
It looks like there's still deadlock potential here, as we invoke the developer's callback while holding the lock.  So if the developer had their own lock and code like (pseudo-code):
```C#
static void Foo()
{
    lock (otherLock)
    {
         NetworkChange.NetworkAvailabilityChanged += Callback;
    }
}

static void Callback()
{
     lock (otherLock)
     {
         ...
     }
}
```
this could deadlock, as Foo will first take a lock on otherlock and then on s_globalLock, whereas the callback will take a lock on s_globalLock and then on otherLock.

I'd need to look in more depth, but it seems like this could actually introduce a deadlock in existing code that didn't previously deadlock due to different locks being used.

Reply:
I have investigated into this deadlock potential scenario, and this should not be a concern.

> as Foo will first take a lock on otherlock and then on s_globalLock, whereas the callback will take a lock on s_globalLock and then on otherLock.

This is not true. Here is the reason:

`Foo()` can grab the `otherLock` first, then the next operation: `NetworkChange.NetworkAvailabilityChanged += Callback;` will grab the `globalLock`. But notice that, for `NetworkChange.NetworkAvailabilityChanged += Callback;` operation:

1. It is the operation to start `AvailabilityChangeListener` & `AddressChangeListener`, then to register any related callbacks with those Listeners. Which means, we should not worry any addressChange or availabilityChange events happen during this operation, because those Listeners haven't started yet & callbacks are not yet registered.
2. This operation will release the `globalLock` as soon as it starts the Listeners & registers their callbacks.

So, after this line `NetworkChange.NetworkAvailabilityChanged += Callback;` 

1. `Foo()` will release the `globalLock`.
2. Any addressChange or availabilityChange events can be detected/handled correctly.

Therefore, when NetworkChange event happens, and user defined `Callback()` get executed, `Foo()` will no longer hold the `globalLock`. The deadlock scenario will never happen.

Reply:
I think that the deadlock situation mentioned by Stephen still exists. Here's a slight modification of Stephens example that I think makes it clearer:
```
static void Foo()
{
    NetworkChange.NetworkAvailabilityChanged += Callback;
    lock (otherLock)
    {
         // Network change occurs, triggering Callback()
         NetworkChange.NetworkAvailabilityChanged += SomeOtherCallback;// Thread 1 here
    }
}

static void Callback()
{
     lock (otherLock)// Thread 2 here
     {
         ...
     }
}
```
In this setup we can end up in a situation where Foo() holds otherLock and is attempting to acquire the NetworkChange static lock, and Callback() holds the NetworkChange global lock but is waiting to acquire otherLock. It looks like this exact case still would have occurred in the old code, but now that the network change lock is static it can happen even with separate instances of NetworkChange.

Reply:
Yeah, the deadlock will happen in multi-thread case. Thanks for pointing that out!

> It looks like there's still deadlock potential here, as we invoke the developer's callback while holding the lock.

I agree with it. I will work on fixing this.